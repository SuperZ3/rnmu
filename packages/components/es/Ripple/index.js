import { Ripple } from "./Ripple.js";
import "react";
import "react-native";
export {
  Ripple
};
