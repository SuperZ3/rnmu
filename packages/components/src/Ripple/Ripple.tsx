import { useState } from 'react';
import {
  Animated,
  ColorValue,
  Easing,
  GestureResponderEvent,
  LayoutChangeEvent,
  Pressable,
  PressableProps,
  StyleSheet,
} from 'react-native';

export interface RippleProps extends PressableProps {
  children: React.ReactNode; // Content of the Ripple
  borderless?: boolean; // Whether to render the ripple outside the view bounds
  rippleColor?: ColorValue; // Color of the ripple effect
  rippleOpcity?: number; // Opacity of th ripple effect
  rippleDuration?: number; // Duration of th ripple effect
  underlayColor?: ColorValue; // Color of the underlay for the highlight effect
  centered?: boolean; // Whether to start the ripple at the center
  disabled?: boolean; // Whether to prevent interaction with the touchable.
}

type RippleElementConfig = {
  uid: number;
  centerX: number;
  centerY: number;
  rippleAnim: Animated.Value;
  R: number;
};

const RippleElement = (props: { rippleConfig: RippleElementConfig }) => {
  const {
    rippleConfig: { centerX, centerY, rippleAnim, R },
  } = props;

  const animatedStyle = {
    top: centerY - R,
    left: centerX - R,
    width: 2 * R,
    height: 2 * R,
    borderRadius: R,
    backgroundColor: 'rgb(0, 0, 0)',
    opacity: 0.3,
    transform: [{ scale: rippleAnim }, { perspective: 1000 }],
  };

  return (
    <Animated.View
      style={[styles.rippleElement, animatedStyle]}
    ></Animated.View>
  );
};

export const Ripple: React.FC<RippleProps> = props => {
  const {
    children,
    borderless,
    rippleColor = 'rgb(0, 0, 0)',
    rippleOpcity = 0.3,
    rippleDuration = 400,
    underlayColor,
    centered = false,
    disabled = false,
    ...rest
  } = props;
  const [target, setTarget] = useState({
    width: 0,
    height: 0,
  });

  const [ripples, setRipples] = useState<RippleElementConfig[]>([]);

  function handleLayout(event: LayoutChangeEvent) {
    const { width, height } = event.nativeEvent.layout;
    setTarget({
      width,
      height,
    });
  }

  function handlePressIn(event: GestureResponderEvent) {
    const { locationX, locationY, timestamp } = event.nativeEvent;
    const halfW = target.width * 0.5;
    const halfH = target.height * 0.5;
    const centerX = centered ? halfW : locationX;
    const centerY = centered ? halfH : locationY;
    const offsetX = Math.abs(centerX - halfW);
    const offsetY = Math.abs(centerY - halfH);
    const R = Math.sqrt((halfW + offsetX) ** 2 + (halfH + offsetY) ** 2);
    const ripple = {
      uid: timestamp,
      centerX,
      centerY,
      rippleAnim: new Animated.Value(0),
      R,
    };

    Animated.timing(ripple.rippleAnim, {
      toValue: 1,
      easing: Easing.inOut(Easing.ease),
      duration: 800,
      useNativeDriver: true,
    }).start();

    setRipples(ripples => ripples.concat(ripple));
  }

  function handlePressOut(event: GestureResponderEvent) {
    const { timestamp } = event.nativeEvent;
    if (ripples.length === 1 && timestamp - ripples[0].uid > 800) {
      setRipples(ripples => ripples.slice(1));
    } else {
      setTimeout(() => {
        setRipples(ripples => ripples.slice(1));
      }, 800);
    }
  }

  return (
    <Pressable
      onPressIn={handlePressIn}
      onPressOut={handlePressOut}
      onLayout={handleLayout}
      style={{
        overflow: 'hidden',
      }}
      {...rest}
    >
      {children}
      {ripples.map(rippleConfig => (
        <RippleElement key={rippleConfig.uid} rippleConfig={rippleConfig} />
      ))}
    </Pressable>
  );
};

export const styles = StyleSheet.create({
  rippleElement: {
    position: 'absolute',
  },
});
