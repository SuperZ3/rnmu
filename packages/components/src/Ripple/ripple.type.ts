import { Key } from 'react';
import { Animated, ColorValue, PressableProps } from 'react-native';

export interface RippleProps extends PressableProps {
  children: PressableProps['children']; // Content of the Ripple
  foreground?: boolean; // Set to true to add the ripple effect to the foreground of the view
  rippleColor?: ColorValue; // Color of the ripple effect
  rippleOpcity?: number; // Opacity of th ripple effect
  rippleDuration?: number; // Duration of th ripple effect
  underlayColor?: ColorValue; // Color of the underlay for the highlight effect
  centered?: boolean; // Whether to start the ripple at the center
  disabled?: boolean; // Whether to prevent interaction with the touchable.
}

export type RippleElementConfig = {
  uid: number;
  centerX: number;
  centerY: number;
  rippleAnim: Animated.Value;
  R: number;
};

export interface RippleElementProps {
  rippleColor: ColorValue;
  rippleOpacity: number;
  isForeground: boolean;
  rippleConfig: RippleElementConfig;
  key: Key;
}
