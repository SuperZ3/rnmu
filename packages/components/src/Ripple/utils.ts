import { Animated, Easing, GestureResponderEvent } from 'react-native';
import { RippleElementConfig } from './ripple.type';

export const RippleElementTestId = 'RippleElementTestId';

export function getRipple(
  event: GestureResponderEvent,
  targetWidth: number,
  targetHeight: number,
  isCentered: boolean,
) {
  const { locationX, locationY, timestamp } = event.nativeEvent;
  const halfW = targetWidth * 0.5;
  const halfH = targetHeight * 0.5;
  const centerX = isCentered ? halfW : locationX;
  const centerY = isCentered ? halfH : locationY;
  const offsetX = Math.abs(centerX - halfW);
  const offsetY = Math.abs(centerY - halfH);
  const R = Math.sqrt((halfW + offsetX) ** 2 + (halfH + offsetY) ** 2);
  return {
    uid: timestamp,
    centerX,
    centerY,
    rippleAnim: new Animated.Value(0),
    R,
  };
}

export function startAnimated(
  ripple: RippleElementConfig,
  duration: number,
  callback?: (finished: boolean) => void,
) {
  Animated.timing(ripple.rippleAnim, {
    toValue: 1,
    easing: Easing.inOut(Easing.ease),
    duration,
    useNativeDriver: true,
  }).start(({ finished }) => callback?.(finished));
}
