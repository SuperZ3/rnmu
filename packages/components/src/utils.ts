import { Platform } from 'react-native';

export function isFunction(v: unknown): v is CallableFunction {
  return v instanceof Function;
}

export const isWeb = Platform.OS === 'web';
