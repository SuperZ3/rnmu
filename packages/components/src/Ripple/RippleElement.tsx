import { Animated, StyleSheet } from 'react-native';
import { RippleElementProps } from './ripple.type';
import { RippleElementTestId } from './utils';

const RippleElement: React.FC<RippleElementProps> = props => {
  const {
    rippleConfig: { centerX, centerY, rippleAnim, R },
    rippleOpacity,
    isForeground,
    rippleColor,
  } = props;

  const animatedStyle = {
    top: centerY - R,
    left: centerX - R,
    width: 2 * R,
    height: 2 * R,
    borderRadius: R,
    backgroundColor: rippleColor,
    opacity: rippleAnim.interpolate({
      inputRange: [0, 1],
      outputRange: [0, rippleOpacity],
    }),
    transform: [{ scale: rippleAnim }, { perspective: 1000 }],
    zIndex: isForeground ? -1 : 0,
  };

  return (
    <Animated.View
      testID={RippleElementTestId}
      style={[styles.rippleElement, animatedStyle]}
    ></Animated.View>
  );
};

export default RippleElement;

export const styles = StyleSheet.create({
  rippleElement: {
    position: 'absolute',
    pointerEvents: 'none',
  },
});
