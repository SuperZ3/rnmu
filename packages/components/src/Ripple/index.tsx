import { isWeb } from '../utils';
import RippleNative from './Ripple';
import RippleWeb from './Ripple.web';
import { RippleProps } from './ripple.type';

export type { RippleProps };

// In react-native-web, the onPress parameter is MouseEvent instead of ResponderEvent
// , causing the uid and locationX deconstructed from event.nativeEvent to be undefined
// in getRipple().
// Therefore, use onPressIn to replace onPress in RippleWeb
// https://necolas.github.io/react-native-web/docs/pressable/
const Ripple = isWeb ? RippleWeb : RippleNative;
export { Ripple };
