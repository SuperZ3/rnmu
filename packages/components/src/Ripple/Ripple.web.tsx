import { useState } from 'react';
import {
  GestureResponderEvent,
  LayoutChangeEvent,
  Pressable,
} from 'react-native';
import { isFunction } from '../utils';
import { getRipple, startAnimated } from './utils';
import { RippleElementConfig, RippleProps } from './ripple.type';
import RippleElement from './RippleElement';

const RippleWeb: React.FC<RippleProps> = props => {
  const {
    children,
    foreground = false,
    rippleColor = 'rgb(0, 0, 0)',
    rippleOpcity = 0.3,
    rippleDuration = 400,
    underlayColor,
    centered = false,
    disabled = false,
    onPressIn,
    onPressOut,
    onLayout,
    style,
    ...rest
  } = props;
  const [target, setTarget] = useState({
    width: 0,
    height: 0,
  });
  const [ripples, setRipples] = useState<RippleElementConfig[]>([]);

  function handleLayout(event: LayoutChangeEvent) {
    const { width, height } = event.nativeEvent.layout;
    setTarget({
      width,
      height,
    });
    onLayout?.(event);
  }

  function handlePressIn(event: GestureResponderEvent) {
    const ripple = getRipple(event, target.width, target.height, centered);
    startAnimated(ripple, rippleDuration);
    setRipples(ripples => ripples.concat(ripple));
    onPressIn?.(event);
  }

  function handlePressOut(event: GestureResponderEvent) {
    const outTimeStemp = event.nativeEvent.timestamp;
    if (ripples.length > 0) {
      // make longPress finish as soon as possible
      const delay =
        ripples.length === 1 && outTimeStemp - ripples[0].uid > rippleDuration
          ? 0
          : rippleDuration;
      setTimeout(() => {
        setRipples(ripples => ripples.slice(1));
      }, delay);
    }
    onPressOut?.(event);
  }

  return (
    <Pressable
      onPressIn={handlePressIn}
      onPressOut={handlePressOut}
      onLayout={handleLayout}
      style={pressed => [
        { overflow: 'hidden' },
        isFunction(style) ? style(pressed) : style,
      ]}
      disabled={disabled}
      {...rest}
    >
      {pressed => {
        return (
          <>
            {isFunction(children) ? children(pressed) : children}
            {ripples.map(rippleConfig => (
              <RippleElement
                key={rippleConfig.uid}
                rippleConfig={rippleConfig}
                rippleOpacity={rippleOpcity}
                rippleColor={rippleColor}
                isForeground={foreground}
              />
            ))}
          </>
        );
      }}
    </Pressable>
  );
};

export default RippleWeb;
