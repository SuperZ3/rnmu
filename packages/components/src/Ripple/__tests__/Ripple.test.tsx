import {
  cleanup,
  fireEvent,
  render,
  screen,
  waitFor,
  waitForElementToBeRemoved,
} from '@testing-library/react-native';
import Ripple from '../Ripple';
import RippleWeb from '../Ripple.web';
import { Platform, Text } from 'react-native';
import { RippleElementTestId } from '../utils';

const EventData = {
  nativeEvent: {
    locationX: 0,
    locationY: 0,
    timestamp: 1,
    layout: {},
  },
};

jest.useFakeTimers();
jest.spyOn(global, 'setTimeout');

describe('Export ripple base on Platform', () => {
  Platform.OS = 'web';
  const comp = jest.mock('../index.tsx');
  expect(comp).toEqual(Ripple);
});

describe('Ripple Component', () => {
  it('should render children correctly', () => {
    const ripple = render(
      <Ripple>
        <Text>Button</Text>
      </Ripple>,
    ).toJSON();

    expect(ripple).toMatchSnapshot();
  });

  describe('ripple effect', () => {
    afterEach(cleanup);
    it('ripple effect props match snapshot', async () => {
      const { toJSON } = render(
        <Ripple centered foreground rippleColor="red" rippleOpcity={1}>
          <Text>Button</Text>
        </Ripple>,
      );
      const comp = await screen.getByText('Button');
      fireEvent(comp, 'press', EventData);
      expect(toJSON()).toMatchSnapshot();
    });

    it('onPress render ripple effect', async () => {
      render(
        <Ripple>
          <Text>Button</Text>
        </Ripple>,
      );
      const comp = await screen.getByText('Button');
      fireEvent(comp, 'press', EventData);
      await waitForElementToBeRemoved(
        () => screen.getByTestId(RippleElementTestId),
        { timeout: 500 },
      );
    });

    it('onLongPress render ripple effect', async () => {
      render(
        <Ripple>
          <Text>Button</Text>
        </Ripple>,
      );
      const comp = await screen.getByText('Button');
      fireEvent(comp, 'longPress', EventData);
      await waitForElementToBeRemoved(
        () => {
          fireEvent(comp, 'pressOut', EventData);
          return screen.getByTestId(RippleElementTestId);
        },
        { timeout: 500 },
      );
    });

    it('rippleDuration prop', async () => {
      render(
        <Ripple rippleDuration={800}>
          <Text>Button</Text>
        </Ripple>,
      );
      const comp = await screen.getByText('Button');
      fireEvent(comp, 'press', EventData);
      await waitFor(
        () => expect(screen.getByTestId(RippleElementTestId)).toBeDefined,
        { timeout: 400 },
      );
      await waitForElementToBeRemoved(
        () => screen.getByTestId(RippleElementTestId),
        { timeout: 800, interval: 400 },
      );
    });
  });

  describe('fire event', () => {
    afterEach(cleanup);
    it('check PressableState', async () => {
      const mockedChildren = jest.fn();
      render(
        <Ripple testID="pressable" testOnly_pressed>
          {mockedChildren}
        </Ripple>,
      );
      const comp = await screen.getByTestId('pressable');
      fireEvent(comp, 'press', EventData);
      expect(mockedChildren).toHaveBeenCalledWith({ pressed: true });
    });

    it.each`
      type
      ${'onLayout'}
      ${'onPress'}
      ${'onLongPress'}
      ${'onPressOut'}
    `('call $type when event fired', async ({ type }) => {
      const mockFn = jest.fn();
      const props = {
        children: <Text>{type}</Text>,
        [type]: mockFn,
      };
      render(<Ripple {...props} />);
      const comp = await screen.getByText(type);
      const event = type
        .substr(2)
        .replace(/^[A-Z]/, (f: string) => f.toLowerCase());
      fireEvent(comp, event, EventData);
      expect(mockFn).toHaveBeenCalledTimes(1);
    });

    it.each`
      type
      ${'onPressIn'}
      ${'onPressOut'}
      ${'onPress'}
      ${'onLongPress'}
    `('not call $type when event fired', async ({ type }) => {
      const mockFn = jest.fn();
      const props = {
        children: <Text>{type}</Text>,
        disabled: true,
        [type]: mockFn,
      };
      render(<Ripple {...props} />);
      const comp = await screen.getByText(type);
      const event = type
        .substr(2)
        .replace(/^[A-Z]/, (f: string) => f.toLowerCase());
      fireEvent(comp, event, EventData);
      expect(mockFn).not.toHaveBeenCalled();
    });
  });
});

describe('RippleWeb Component', () => {
  const Ripple = RippleWeb;
  it('should render children correctly', () => {
    const { getByText } = render(
      <Ripple>
        <Text>Button</Text>
      </Ripple>,
    );

    expect(getByText('Button')).toBeTruthy();
  });

  describe('ripple effect', () => {
    afterEach(cleanup);
    it('ripple effect props match snapshot', async () => {
      const { toJSON } = render(
        <Ripple centered foreground rippleColor="red" rippleOpcity={1}>
          <Text>Button</Text>
        </Ripple>,
      );
      const comp = await screen.getByText('Button');
      fireEvent(comp, 'pressIn', EventData);
      expect(toJSON()).toMatchSnapshot();
    });

    it('onPressIn render ripple effect', async () => {
      render(
        <Ripple>
          <Text>Button</Text>
        </Ripple>,
      );
      const comp = await screen.getByText('Button');
      fireEvent(comp, 'pressIn', EventData);
      expect(screen.getByTestId(RippleElementTestId)).toBeDefined();
      fireEvent(comp, 'pressOut', EventData);
      await waitForElementToBeRemoved(() =>
        screen.getByTestId(RippleElementTestId),
      );
    });
  });

  describe('fire event', () => {
    afterEach(cleanup);
    it('check PressableState', async () => {
      const mockedChildren = jest.fn();
      render(
        <Ripple testID="pressable" testOnly_pressed>
          {mockedChildren}
        </Ripple>,
      );
      const comp = await screen.getByTestId('pressable');
      fireEvent(comp, 'pressIn', EventData);
      expect(mockedChildren).toHaveBeenCalledWith({ pressed: true });
    });

    it.each`
      type
      ${'onLayout'}
      ${'onPressIn'}
      ${'onPressOut'}
    `('call $type when event fired', async ({ type }) => {
      const mockFn = jest.fn();
      const props = {
        children: <Text>{type}</Text>,
        [type]: mockFn,
      };
      render(<Ripple {...props} />);
      const comp = await screen.getByText(type);
      const event = type
        .substr(2)
        .replace(/^[A-Z]/, (f: string) => f.toLowerCase());
      fireEvent(comp, event, EventData);
      expect(mockFn).toHaveBeenCalledTimes(1);
    });

    it.each`
      type
      ${'onPressIn'}
      ${'onPressOut'}
    `('not call $type when event fired', async ({ type }) => {
      const mockFn = jest.fn();
      const props = {
        children: <Text>{type}</Text>,
        disabled: true,
        [type]: mockFn,
      };
      render(<Ripple {...props} />);
      const comp = await screen.getByText(type);
      const event = type
        .substr(2)
        .replace(/^[A-Z]/, (f: string) => f.toLowerCase());
      fireEvent(comp, event, EventData);
      expect(mockFn).not.toHaveBeenCalled();
    });
  });
});
