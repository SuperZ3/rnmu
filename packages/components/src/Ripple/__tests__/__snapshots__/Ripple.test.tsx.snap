// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ripple Component ripple effect ripple effect prop match snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onLayout={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "overflow": "hidden",
      },
      undefined,
    ]
  }
>
  <Text>
    Button
  </Text>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "red",
        "borderRadius": 0,
        "height": 0,
        "left": 0,
        "opacity": 0,
        "pointerEvents": "none",
        "position": "absolute",
        "top": 0,
        "transform": [
          {
            "scale": 0,
          },
          {
            "perspective": 1000,
          },
        ],
        "width": 0,
        "zIndex": -1,
      }
    }
    testID="RippleElementTestId"
  />
</View>
`;

exports[`Ripple Component ripple effect ripple effect props match snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onLayout={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "overflow": "hidden",
      },
      undefined,
    ]
  }
>
  <Text>
    Button
  </Text>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "red",
        "borderRadius": 0,
        "height": 0,
        "left": 0,
        "opacity": 0,
        "pointerEvents": "none",
        "position": "absolute",
        "top": 0,
        "transform": [
          {
            "scale": 0,
          },
          {
            "perspective": 1000,
          },
        ],
        "width": 0,
        "zIndex": -1,
      }
    }
    testID="RippleElementTestId"
  />
</View>
`;

exports[`Ripple Component should render children correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onLayout={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "overflow": "hidden",
      },
      undefined,
    ]
  }
>
  <Text>
    Button
  </Text>
</View>
`;

exports[`RippleWeb Component ripple effect ripple effect props match snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onLayout={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "overflow": "hidden",
      },
      undefined,
    ]
  }
>
  <Text>
    Button
  </Text>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "red",
        "borderRadius": 0,
        "height": 0,
        "left": 0,
        "opacity": 0,
        "pointerEvents": "none",
        "position": "absolute",
        "top": 0,
        "transform": [
          {
            "scale": 0,
          },
          {
            "perspective": 1000,
          },
        ],
        "width": 0,
        "zIndex": -1,
      }
    }
    testID="RippleElementTestId"
  />
</View>
`;
